tempesta_fw/sock.c

```diff
@@ -78,6 +85,63 @@ skb_sender_cpu_clear(struct sk_buff *skb)
 #endif
 }
 
+/**
+ * Enters critical section synchronized with ss_synchronize().
+ * Active networking operations which involves SS callback calls must be
+ * protected by the guard: don't enter the section if the system is about
+ * to shutdown. The only exception is closing activity - this is the only
+ * activity allowed in progress of shutdown process.
+ *
+ * Returns zero (SS_OK) if we're in critical section and SS_BAD if shutdown
+ * process in progress and we can't enter the section.
+ */
+static int
+__active_guard_enter(unsigned long val, int cpu)
+{
+	atomic64_t *active = per_cpu_ptr(&__ss_active, cpu);
+	unsigned long a_new, a_old = atomic64_read(active);
+
+	while (1) {
+		if (unlikely(a_old >= SS_STOP_VAL))
+			return SS_BAD;
+
+		a_new = a_old + val;
+		a_new = atomic64_cmpxchg(active, a_old, a_new);
+		if (likely(a_new == a_old))
+			return SS_OK;
+		a_old = a_new;
+	}
+}
+
+static void
+__active_guard_exit(unsigned long val, int cpu)
+{
+	atomic64_sub(val, per_cpu_ptr(&__ss_active, cpu));
+}
+
+#define ss_active_guard_enter(v)	__active_guard_enter(v,		\
```

Comment:
Actually, guard is almost synonym for `lock`, so there could be many guards. `SS_CALL_GUARD_*` also explicitly says that the guard is for `SS_CALL()`, SS upcall interface.